### PR TITLE
#51 free liquidity -- Draft PR

### PR DESCRIPTION
**Pull request by blewater**
Please note, changesets from previous open PRs have crept in.
### A. Draft PR for soliciting feedback on completed 1..4 issue items.

> 1. Make the gas associated with all market messages identical, so that MsgCancelLimitOrder potentially carries same cost as MsgCancelReplaceLimitOrder.
> 2. New orders that add liquidity should receive a gas rebate proportional to how much of the order was filled already (0% fill means 100% rebate):
> https://github.com/e-money/em-ledger/blob/12b0302ca45100ddeaebfd882392b4731ff16837/x/market/keeper/keeper.go#L334
> 3. Canceled orders always pay full gas.
> 4. Replaced orders are only eligible for the gas rebate in 2. if they are replacing an order older than, say, 5 minutes. This requires #34 to work.

### B. This PR does not address issue section: ```Additional considerations```. 
Suggestions below

> It must not be possible to exploit the rebate mechanism in a way where it becomes cheaper (or even free) to remove liquidity, i.e. crafting an order to get almost filled but adding a small amount to the book should not result in a rebate.
> It must not be possible to spam the chain with useless transactions. Only behaviour that is net adding liquidity

#### Half baked ideas

1. Make Liquidity Fee a minimal dust amount i.e. 250 resulting in *free* single transactions, but costly for sending a large number of transactions. A known practice.
2. Do not rebate more than x number of *free* liquid orders per account at any time. *Not a great solution in isolation.*
3. Do not rebate more than x number of duplicate instruments orders per account at any time.
4. Do not rebate any duplicate instrument order that does not offer qty/pricing spread differential above a threshold i.e. the moving average of the last x booked instruments orders + 15%.

### C. Three Market parameters have been added

- May leave them like that (beginblock)
- Add them to genesis
- Const them (what if we change them?)

```Proto
message TxParams {
  // default fee for a market transaction.
  uint64 trx_fee = 1;

  // Valid liquidity adding orders are free(0) or set to this minimal fee
  uint64 liquid_trx_fee = 2;

  // Minutes interval for eligible replacing transactions to receive a rebate.
  // For a rebate to apply, the replacing transaction should occur these minutes
  // after the signer's original trx.
  int64 liquidity_rebate_minutes_span = 3;
}
```

### CODE REVIEW

**Review by haasted — CHANGES_REQUESTED** [review]

File: x/market/keeper/keeper.go [hunk 1]
```diff
@@ -115,41 +127,154 @@ func (k *Keeper) createExecutionPlan(ctx sdk.Context, SourceDenom, DestinationDe
...
+			*callerErr = handlePanic(sdkErr)
 		}
-
-		commitTrade()
 	}()
 
+	k.postOrderSpendGas(ctx, order, orderGasMeter, callerErr)
+
+	// Roll back any state changes made by the aggressive FillOrKill order.
+	if order.TimeInForce == types.TimeInForce_FillOrKill {
```
[the remark this answers is not in the document]
Comment (blewater) [reply x/market/keeper/keeper.go:261]: thank you Sir!

File: x/market/keeper/keeper.go [hunk 2]
```diff
@@ -115,41 +127,154 @@ func (k *Keeper) createExecutionPlan(ctx sdk.Context, SourceDenom, DestinationDe
...
+}
+
+// postNewOrderSingle is a deferred function for NewOrderSingle. It spends gas
+// on any order and adds non fill-or-kill orders to the order-book.
+func (k *Keeper) postNewOrderSingle(
+	ctx sdk.Context, orderGasMeter sdk.GasMeter, order *types.Order,
+	commitTrade func(), callerErr *error,
+) {
+	// Catch NewSingleOrder() panics
+	if orderErr := recover(); orderErr != nil {
```
[the remark this answers is not in the document]
Comment (blewater) [reply x/market/keeper/keeper.go]: Removed recover()

File: x/market/types/market.pb.go [hunk 3]
```diff
@@ -123,6 +123,9 @@ type Order struct {
 	Destination       types.Coin                             `protobuf:"bytes,8,opt,name=destination,proto3" json:"destination" yaml:"destination"`
 	DestinationFilled github_com_cosmos_cosmos_sdk_types.Int `protobuf:"bytes,9,opt,name=destination_filled,json=destinationFilled,proto3,customtype=github.com/cosmos/cosmos-sdk/types.Int" json:"destination_filled" yaml:"destination_filled"`
 	Created           time.Time                              `protobuf:"bytes,10,opt,name=created,proto3,stdtime" json:"created" yaml:"created"`
+	// applicable to CancelReplace orders: the creation timestamp of the original
+	// order.
+	OrigOrderCreated time.Time `protobuf:"bytes,11,opt,name=orig_order_created,json=origOrderCreated,proto3,stdtime" json:"orig_order_created" yaml:"orig_order_created"`
```
[the remark this answers is not in the document]
Comment (blewater) [reply x/market/types/market.pb.go]: Removed the field.

File: x/market/keeper/keeper.go [hunk 4]
```diff
@@ -115,41 +127,154 @@ func (k *Keeper) createExecutionPlan(ctx sdk.Context, SourceDenom, DestinationDe
...
+	payPct := dstFilled.ToDec().Quo(dstAmount.ToDec())
+	payGas := payPct.Mul(totalRebate.ToDec()).RoundInt().Uint64()
+
+	return payGas
+}
+
+// calcReplaceOrderGas computes gas for replacement orders. Because it needs
+// the original order's creation time, this func acts is a precondition for
+// calcOrderGas().
+func (k *Keeper) calcReplaceOrderGas(
```
[the remark this answers is not in the document]
Comment (blewater) [reply x/market/keeper/keeper.go:185]: Done

**Comment by blewater** [comment]
### Freezing this feature while a better SDK or an alternative solution is found.
Accomplished in this work are predictable gas costs for market orders **within the market module**.
However, before entering the market order, RunTx() spends gas outside within the Ante handlers chain, which in the DeliverTx tests they ramp up to ~55000. These can be zeroed out with an extra ante handler for the market module but at the severe disadvantage of exposing a security vulnerability of gas-free transactions when the market module panics.

A solution explored for mitigating the security vulnerability is adding a BaseApp custom panic handler, combined with the handler having access to the transaction gas meter for charging a gas fee which sadly it does not. As a future consideration, we could open an issue with the Cosmos SDK team to add gas meter access to panic-handlers applicable to this and other business functionality.

**Comment by blewater** [comment]
There is a completed Cosmos SDK PR that meets the needs of this functionality, and it was stalled because of a lack of an accompanied ADR. We could pick up the pieces for this and contribute the ADR and facilitate the re-opening of the stalled PR to complete this [add gas refund handler to refund the remaining gas fee at the end of transaction #8993](https://github.com/cosmos/cosmos-sdk/pull/8993).

**Pull request closed without merging**